qiskit/optimization/problems/quadratic_program.py

```diff
@@ -151,36 +151,102 @@ def _add_variable(self,
+        if var_count < 1:
+            raise QiskitOptimizationError(
+                "Cannot create non-positive number of variables: {}".format(var_count))
+        if not name and formatter:
+            name = 'x'
+        if name and not formatter:
+            if var_count > 1:
+                formatter = '{}'
+            else:
+                formatter = ''
+                if name in self._variables_index:
+                    raise QiskitOptimizationError("Variable name already exists: {}".format(name))
+        if not name and not formatter:
+            name = 'x'
+            formatter = '{}'
+        return self._create_var_dict_update_maps(lowerbound, upperbound, vartype, name, formatter,
+                                                 var_count)
+
+    def _create_var_dict_update_maps(self,
+                                     lowerbound: Union[float, int],
+                                     upperbound: Union[float, int],
+                                     vartype: VarType,
+                                     name: str,
+                                     formatter: str,
+                                     var_count: int) -> Dict[str, Variable]:
+        if '{{}}' in formatter:
+            raise QiskitOptimizationError(
+                "Formatter cannot contain nested substitutions: {}".format(formatter))
+        substitution_count = formatter.count('{}')
+        if substitution_count > 1:
+            raise QiskitOptimizationError(
+                "Formatter cannot contain more than one substitution: {}".format(formatter))
+        new_var_dict = {}
+        k = self.get_num_vars()
+        for _ in range(var_count):
+            while name + formatter.format(k) in self._variables_index:
+                k += 1
+            indexed_name = name + formatter.format(k)
+            new_var_dict[indexed_name] = self._create_var_update_maps(lowerbound,
+                                                                      upperbound,
+                                                                      vartype, indexed_name)
+        return new_var_dict
+
+    def _create_var_update_maps(self,
+                                lowerbound: Union[float, int],
+                                upperbound: Union[float, int],
+                                vartype: VarType,
+                                name: str) -> Variable:
+        self.variables_index[name] = len(self.variables)
+        variable = Variable(self, name, lowerbound, upperbound, vartype)
+        self.variables.append(variable)
+        return variable
+
+    def var_dict(self,
+                 lowerbound: Union[float, int] = 0,
+                 upperbound: Union[float, int] = INFINITY,
+                 vartype: VarType = VarType.CONTINUOUS,
+                 name: Optional[str] = None,
+                 formatter: Optional[str] = None,
+                 var_count: int = 1) -> Dict[str, Variable]:
```

Comment:
It would be better if `var_count` behaves as `keys` of docplex. `keys` is `Union[int, Sequence]` and it is converted to a sequence if it is an integer.
http://ibmdecisionoptimization.github.io/docplex-doc/mp/docplex.mp.model.html#docplex.mp.model.Model.binary_var_dict

Reply:
We can think of a use case such as `quadratic_program.binary_var_dict(name='x', keys=range(5, 10))`.

Reply:
Currently, `quadratic_program` selects the next `var_count` free variables with prefix `name`, so the indices are not necessarily consecutive. This avoids attempting to bind variables that have already been bound in `quadratic_program`. Docplex basically just converts each element of the sequence to a string and appends it to `name`. Should the function you're requesting simply throw an error if (for example), one of the `x_5, x_6, ..., x_7` is already bound?

Reply:
Yes, I think it makes sense to raise an error if `x_5` is defined. If a variable name is explicitly given and it collides with an existing variable, we need to raise an error. If a variable name is not explicitly given, we need to look for an index that does not collide with existing variables. Does it make sense?

Reply:
`keys: int` is interpreted as the number of variables to create. It makes sense for `keys: int` to indicate that the next `keys` variables are chosen non-consecutively. So `keys: int` will not throw an error, but `keys: Sequence` might. That's sensible to me.